docs/How_To/Tutorials/add_citation.rst

```diff
@@ -0,0 +1,161 @@
+        self.k = k
+
+        # load the pdf and create the vector store
+        self.chroma = Chroma(embedding_function = embedding_function)
+        documents = PDFMinerLoader(self.filename).load()
+        documents = SentenceTransformersTokenTextSplitter(chunk_overlap=0).split_documents(documents)
+
+        # This is the new code to add the source to the metadata
+        for i in range(len(documents)):
+            documents[i].metadata["chunk_id"] = f"chunk_{i}"
+        # End of the new code
+
+        logger.info(f"Adding {len(documents)} documents to the vector store")
+        self.chroma.add_documents(documents)
+        logger.info("Finished adding documents to the vector store")
+
+Next, when we execute the search, we will add the resources using the `add_resources` method so that later the `CitationValidation` can aware of these resources.
+
+.. code-block:: python
+
+    def execute(self, query):
+        """
+        Execute the action by searching the document store for the query
+
+        Args:
+            query (str): The query to search for
+
+        Returns:
+            str: The search results combined into a single string
+        """
+
+        results = self.chroma.search(query, search_type="mmr", k=self.k)
+
+        # This is the new code to add the resources
+        resources = [
+            {"Document": result.page_content, "Source": result.metadata["chunk_id"]}
+            for result in results
+        ]
+        self.add_resources(resources)
+        # End of the new code
+
+        return "\n\n".join([result.page_content for result in results])
+
+We are done! Again, to test the citation validation, let's remove the Google search action from the agent and run the agent. Ask a question that will trigger the document search action. The agent should return the response with the citation.
+
+.. code-block:: yaml
+
+    qa_agent:
+        ...
+        actions:
+            - ${doc_search}
+            # - ${google_search}
+
+
+.. code-block:: bash
+
+    Ask me a question: What is data leakage
+    2024-05-09 00:24:57.552 | INFO     | sherpa_ai.agents.base:run:70 - Action selected: ('DocumentSearch', {'query': 'What is data leakage'})
+    Data leakage refers to the potential for data to be unintentionally exposed or disclosed to unauthorized parties [1](doc:chunk_5), [3](doc:chunk_45). In the context provided, data leakage is discussed in relation to the presence of inter-dataset code duplication and the implications for the evaluation of language models in software engineering research [1](doc:chunk_5). It is highlighted as a potential threat that researchers need to consider when working with pre-training and fine-tuning datasets for language models [1](doc:chunk_5). By acknowledging the risk of data leakage due to code duplication, researchers can enhance the robustness of their evaluation methodologies and improve the validity of their results [1](doc:chunk_5).
```

Comment:
"chunk_5" etc is what you're calling "source" above? does that need to be a unique ID? in some cases I'm assuming that we would want to allow the front end to show the text of the chunk when clicked on this link

Reply:
I added instructions to output the chunk table so that one can check the chunk associated with the chunk id